core/src/main/java/io/grpc/BinaryLogProvider.java

```diff
@@ -14,66 +14,34 @@
  * limitations under the License.
  */
 
-package io.grpc.internal;
+package io.grpc;
 
 import com.google.common.annotations.VisibleForTesting;
-import io.grpc.CallOptions;
-import io.grpc.Channel;
-import io.grpc.ClientCall;
-import io.grpc.ClientInterceptor;
-import io.grpc.ClientInterceptors;
-import io.grpc.InternalClientInterceptors;
-import io.grpc.InternalServerInterceptors;
-import io.grpc.InternalServiceProviders;
-import io.grpc.InternalServiceProviders.PriorityAccessor;
-import io.grpc.MethodDescriptor;
+import com.google.common.base.Preconditions;
 import io.grpc.MethodDescriptor.Marshaller;
-import io.grpc.ServerCallHandler;
-import io.grpc.ServerInterceptor;
-import io.grpc.ServerMethodDefinition;
 import io.opencensus.trace.Span;
 import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.Closeable;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.OutputStream;
 import java.nio.ByteBuffer;
-import java.util.Collections;
 import java.util.logging.Logger;
 import javax.annotation.Nullable;
 
+// TODO(zpencer): rename class to AbstractBinaryLog
+@Internal
 public abstract class BinaryLogProvider implements Closeable {
```

Comment:
It doesn't seem this needs to be in io.grpc. Only the two public methods are necessary. Those can be abstract in a separate class from BinaryLog (like is done now) or they can be on BinaryLog itself. Long-term I'm not sure it makes sense for this to be locked down.

Reply:
I wanted something that can be passed into ManagedChannelBuilder and ServerBuilder, and I remember io.grpc can not depend on io.grpc.internal. Or are you suggesting to nest it inside of io.grpc.BinaryLog ?

Reply:
I'm suggesting to strip the body of BinaryLogProvider out. The two-method interface is all that's necessary in io.grpc. I'd understand a bit more if the fact that those two methods are `final` limited the scope of the API, but they don't really limit the power of the interface.